### PR TITLE
Fail travis on clang-format or build warnings

### LINKED ISSUE #93: Add GitHub PR templates and hooks
See dicussion at https://3.basecamp.com/3692715/buckets/3007033/messages/507250988

- Add a PR template with a checklist: 
    - docs
    - tests 
    - format

We can add a hook that runs or checks clang-format -- or add the check to travis. Can also have travis fail on compiler warnings.

Also, we can add an issue template (e.g. https://github.com/opencv/opencv/issues/new).

### PR DESCRIPTION
**Pull request by leokoppel**
See #93 
Related: #101 

- Add separate format.bash
    - Separate script can be run by regular users for convenience
    - Format files only in ./wave_*/ - (avoid any generated in build directories)
- Fail build on clang-format replacements
    - If failed, show the affected files
- Fail build on warnings
    - Add -Werror to fail on warnings.
    - Make an exception for deprecated-declarations warnings
- Move CI scripts to scripts/ci directory
- Delete sync_docs.bash
    Should have been removed in documentation PR

### CODE REVIEW

**Comment by leokoppel** [comment]
Here is an example of what it looks like when the check fails: https://travis-ci.org/wavelab/libwave/builds/236159078

> The following files were changed after running clang-format:
 wave_vision/include/wave/vision/fast_detector.hpp  | 27 +++++++++++-----------
 .../include/wave/vision/feature_detector.hpp       | 11 +++++----
 wave_vision/src/fast_detector.cpp                  | 26 ++++++++++-----------
 wave_vision/tests/fast_tests.cpp                   | 18 +++++++++------
 4 files changed, 43 insertions(+), 39 deletions(-)
clang-format check failed

**Comment by chutsu** [comment]
@leokoppel 

> Which is more valuable: run the whole build and unittests and check clang-format, or check format right away and immediately fail the build?

I would opt for the second option "check format right away and immediately fail the build"

File: scripts/format.bash [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# This variable can be set to use a different version of clang-format
+: ${FORMAT_EXE:=clang-format}
+
+# Format all source files. Run from the source root directory.
+find ./wave_*/ -name "*.hpp" -o -name "*.cpp" | xargs ${FORMAT_EXE} -i
```
Comment (chutsu) [inline scripts/format.bash]: maybe have `./wave_*/` as `$1` bash script argument or variable?

Comment (leokoppel) [inline scripts/format.bash]: OK, it is now only the default and you can pass another path.

File: wave_containers/tests/containers/landmark_measurement_test.cpp [hunk 2]
```diff
@@ -35,13 +35,13 @@ TEST(LandmarkContainer, insert) {
 TEST(LandmarkContainer, emplace) {
     auto m = TestContainer{};
     auto now = std::chrono::steady_clock::now();
-    auto res = m.emplace(now, CameraSensors::Left, 1, Vec2{1.2, 3.4});
+    auto res = m.emplace(now, CameraSensors::Left, 1ul, Vec2{1.2, 3.4});
```
Comment (chutsu) [inline wave_containers/tests/containers/landmark_measurement_test.cpp]: just a `u` should suffice? `1u`, I think its more readable the `l` looks too much like 1.

File: scripts/format.bash [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+# Usage: format.bash [path to search]
+# clang-format will be called for each .hpp and .cpp file in the given path(s)
+# The default is: ./wave_*/
+
+SEARCH_PATHS=${@:-./wave_*/}
+
+# This variable can be set to use a different version of clang-format
+: ${FORMAT_EXE:=clang-format}
```
Comment (chutsu) [inline scripts/format.bash]: What does the `:` do again? make add a short comment :D

File: wave_containers/tests/containers/landmark_measurement_test.cpp (hunk 2 above)
Comment (leokoppel) [inline wave_containers/tests/containers/landmark_measurement_test.cpp]: Sure.

File: scripts/format.bash (hunk 3 above)
Comment (leokoppel) [inline scripts/format.bash]: https://www.google.ca/search?q=bash+colon

**Pull request merged**